authenticate/handlers.go

```diff
@@ -97,7 +97,13 @@ func (a *Authenticate) SignIn(w http.ResponseWriter, r *http.Request) {
 	// original `state` parameter received from the proxy application.
 	state := r.Form.Get("state")
 	if state == "" {
-		httputil.ErrorResponse(w, r, "no state parameter supplied", http.StatusForbidden)
+		httputil.ErrorResponse(w, r, "no state parameter supplied", http.StatusBadRequest)
+		return
+	}
+
+	redirectURI := r.Form.Get("redirect_uri")
+	if redirectURI == "" {
+		httputil.ErrorResponse(w, r, "no redirect_uri parameter supplied", http.StatusBadRequest)
 		return
 	}
 
```

Comment:
I think we can avoid the duplicate `r.Form.Get` call on line 110 and use `redirectURI`

```go
	redirectURL, err := url.Parse(redirectURI)
```